### ISSUE TITLE
Change Captive-Portal pages with bootstrap

### ISSUE THREAD

**Issue by tobiaskastl** [openNDS/openNDS#112]
Hey!

Now that my captive portal works, I would like to change the layout and colors (if that's allowed & possible).

I would like to "disable" the step to accept the "terms of service". 

I want to inform the user like this: "By clicking _Continue_, you automatically accept the terms of service." Then a button with "Continue" and a second underneath with "Read terms of service". 

If possible - where do I need to change this? In the php-files of the folder "forward_authentication_service"?

I managed to run nodogsplash with bootstrap, but only because there were html-files I could edit. I am not very familiar with PHP, that's why I'm asking if implementing bootstrap or changing the layout at all is possible.

**Comment by bluewavenet** [comment]
@tobiaskastl 

> with bootstrap,

What is bootstrap?

You are most likely running with the default:
`login_option_enabled 1`

The html served as the "splash page" is generated dynamically by the default login script.
This is a bash script residing at /etc/opennds/login.sh

It is written to be fairly easy to follow the logic and will be very easy to modify the html.
The file /etc/opennds/htdocs/splash.css is used and can also be modified as required.

Look carefully at these two files and, hopefully, all will be revealed :-)

**Comment by bluewavenet** [comment]
@tobiaskastl 
OK, I looked it up. I can't see the point to be honest. 
Note also Bootstrap is heavily dependent on an "on demand" download of javascript files. 
Captive Portal Detection (CPD) is the process that is supplied by all operating systems to enable the popup splash page from a Captive Portal. For security reasons, downloading of Portal files including javascript is forbidden by most implementations of CPD.

**Comment by tobiaskastl** [comment]
@bluewavenet 

The only "login.sh"-file I can find is in _openNDS-8.0.0/forward_authentication_service/PreAuth_.
But I have three php-files in /etc/opennds - "fas-aes", "fas-aes-https" and "fas-hid".

For example, I just edited the following:

![image](https://user-images.githubusercontent.com/72079975/107978046-6d1bb800-6fbc-11eb-8aba-db0f8769dcfb.png)

This did not change anything on my captive portal.

> You are most likely running with the default:
> `login_option_enabled 1`

Not really - I had the default (0). After adding this line, also nothing changed.

**Comment by tobiaskastl** [comment]
I would like to just skip the splash page that has the two buttons "Accept Terms of Service" and "Read Terms of Service", but I don't know how. I just tested "login_option_enabled 2" for the mail-authentication; this worked.

**Comment by mk248** [comment]
This use case might be better served with nodogsplash, which is cut down to support simple use cases (click to connect) via the legacy templated pages.  

In the code you quoted, change the form action to "opennds_auth" instead of "opennds_preauth"-- and it should authenticate immediately upon clicking the submit button.  The other pages won't be called.  You could rename the button value "Accept" instead of "Continue".  It would be good to change something like that visible on the page to confirm that this code is really running instead of the legacy page.

**Comment by bluewavenet** [comment]
@tobiaskastl 
I could pretend I did not tell you where the installed login.sh deliberately as a test... but in actual fact I did put it in but incorrectly :-)

It should have read:

 The html served as the "splash page" is generated dynamically by the default login script.
 This is a bash script residing at **/usr/lib/opennds/login.sh**

If you want the simple click to continue page, then both modes 0 and 1 will give it to you, but mode 0 is special as it enables FAS if configured, defaulting to login.sh if not.

You will see in login.sh that it is simple to change the html that is served for each page of login, or even change the number of pages.

The example edit you showed above indicates you are along the right lines.

**Comment by tobiaskastl** [comment]
@bluewavenet Thanks, the file in /usr/lib is the right one!

@mk248 I changed the form action to "opennds_auth" but I get the error "connection refused" after clicking on the button.

**Comment by bluewavenet** [comment]
@mk248 

> This use case might be better served with nodogsplash

Unfortunately not, as the requirement is specifically to allow access to 2 separate web servers residing on the openNDS device ONLY after the client has authenticated (ie access is blocked until authenticated). Presumably at a later date a secure login will be implemented (otherwise it is all a little pointless) - currently this is, as far as I am aware, a practical but nevertheless academic exercise.

> change the form action to "opennds_auth" instead of "opennds_preauth"-- and it should authenticate immediately upon clicking the submit button on this first page

This will not work as login.sh uses the do_ndsctl API call to authenticate. This is called from the landing page as can be seen from the script. It _could_ be changed as you suggest but you would also have to construct a suitable query string including the rhid - there is no advantage in doing this.

Note, the legacy templated splash.html page is deprecated and will be removed at the next major release.

**Comment by tobiaskastl** [comment]
@bluewavenet 
You described my situation pretty accurate.

Isn't it somehow possible to just link the "Accept Terms of Service"-button with the landing page and include the authentication in this step? If that's too laborious, I will continue with the extra page - doesn't matter.

**Comment by bluewavenet** [comment]
@tobiaskastl 
Hint - find this section of login.sh (near the end):
```
# Check login mode and serve appropriate form, mode 0 is the default if FAS is disabled
if [ $mode -eq 0 ]; then
	click_to_continue

elif [ $mode -eq 1 ]; then
	click_to_continue

elif [ $mode -eq 2 ]; then
	name_email_login

else
	echo "<b>Invalid login mode #:$mode</b>"
	footer
fi

```

and add a new mode like this:
```
# Check login mode and serve appropriate form, mode 0 is the default if FAS is disabled
if [ $mode -eq 0 ]; then
	click_to_continue

elif [ $mode -eq 1 ]; then
	click_to_continue

elif [ $mode -eq 2 ]; then
	name_email_login

elif [ $mode -eq 3 ]; then
	thankyou_page

else
	echo "<b>Invalid login mode #:$mode</b>"
	footer
fi

```
Now update the config file in the relevant place ;-)
And restart openNDS.

Now try connecting.
This won't be exactly what you want but will be pretty close - at least for a "hint" :-D

**Comment by mk248** [comment]
OK yes I see that you can't just jump to opennds_auth.  A very minimal hack would be to make the button on the first page send "landing=yes" (instead of "continue=clicked") which would shortcut to the third page.  landing=yes  is what the button on the second page does.

Main reason I'm here though is that I've found a security issue that I'd like to talk to @bluewavenet in private but not sure how to do that.

**Comment by bluewavenet** [comment]
@mk248 
If your reason for being here is not related to "Change Captive-Portal pages with bootstrap", please open a new issue.

**Comment by tobiaskastl** [comment]
@bluewavenet Awesome, thanks! 

One quick question: after clicking "Accept and Authenticate" on my _thankyou-page_, I get authenticated and land on the _landing-page_. 

Would it be possible to add a _LogOut-button_ at the bottom of the landing-page? I can't find a method in "/usr/lib/opennds/login.sh" that calls the content of "status.page" (because there is in fact a logout-button - I thought I could copy it from there). Would be awesome if you could answer me this!

**Comment by bluewavenet** [comment]
@tobiaskastl 
Try adding:
```
<form action=\"/opennds_deny/\" method=\"get\">
	<input type=\"submit\" value=\"Logout\" >
</form>
```

**Comment by tobiaskastl** [comment]
@bluewavenet Awesome! Works perfect :-) 

I changed my style to this:

![image](https://user-images.githubusercontent.com/72079975/108125756-38395f00-70a9-11eb-976a-cd76f960ce2b.png)

Any suggestions for improvement on the style?

**Comment by bluewavenet** [comment]
Looks good!

**Comment by tobiaskastl** [comment]
@bluewavenet Thanks!

I would like to add a second picture in the footer. What would be the right path for "src", if the picture is called "Logo.png" and is in "/etc/opennds/htdocs/images/Logo.png"?
